### PR TITLE
Build Columnar.so and make Citus depends on it 

### PR DESCRIPTION
**Pull request by Quicheche**
Separate Columnar & Citus step 1: generate columnar.so and make Citus depends on it.  We will build and install Columnar first, then build Citus by using columnar.so as lib. Changes mainly involves:

1. New Columnar Makefile
2. Mary's commit to create symbolic link in /columnar for safelib
3. Build columnar first and use columnar.so when building citus

### CODE REVIEW

File: src/backend/columnar/Makefile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+       columnar_tableam.o \
+	   columnar_storage.o \
+	   columnar_compression.o \
+	   columnar_customscan.o \
+	   columnar_debug.o \
+	   columnar_metadata.o \
+	   columnar_reader.o \
+	   columnar_writer.o \
+	   mod.o \
+	   write_state_management.o
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:20]: Consider a `pathsubst` expression like in `distributed/Makefile`.

File: src/backend/columnar/Makefile [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	   columnar_debug.o \
+	   columnar_metadata.o \
+	   columnar_reader.o \
+	   columnar_writer.o \
+	   mod.o \
+	   write_state_management.o 
+
+PG_CPPFLAGS += -I$(libpq_srcdir) -I$(safestringlib_srcdir)/include 
+
+DATA = $(wildcard sql/*--*.sql)
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:24]: Is this already being installed as part of the `generated_sql_files` in `distributed/Makefile`?

File: src/backend/columnar/Makefile [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+REGRESS = columnar
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:25]: I don't think we will be running the tests from here, right? So I don't think we need this.

File: src/backend/columnar/columnar_tableam.c [hunk 4]
```diff
@@ -1026,12 +1026,12 @@ columnar_vacuum_rel(Relation rel, VacuumParams *params,
...
-		 * erroring during auto-vacuum.
-		 */
-		return;
-	}
+	 	/*
+	 	 * Skip if the extension catalogs are not up-to-date, but avoid
+	 	 * erroring during auto-vacuum.
+	 	 */
+	 	return;
+	}
```
Comment (jeff-davis) [inline src/backend/columnar/columnar_tableam.c:1034]: Unnecessary diff.

File: src/backend/distributed/Makefile [hunk 5]
```diff
@@ -34,14 +32,16 @@ OBJS += \
 	$(patsubst $(citus_abs_srcdir)/%.c,%.o,$(foreach dir,$(SUBDIRS), $(sort $(wildcard $(citus_abs_srcdir)/$(dir)/*.c))))
 
 # be explicit about the default target
-all:
+all:
```
Comment (jeff-davis) [inline src/backend/distributed/Makefile:35]: unnecessary diff

File: src/backend/distributed/Makefile [hunk 6]
```diff
@@ -86,7 +86,7 @@ endif
 cleanup-before-install:
 	rm -f $(DESTDIR)$(datadir)/$(datamoduledir)/citus*
 
-install: cleanup-before-install
+install: cleanup-before-install
```
Comment (jeff-davis) [inline src/backend/distributed/Makefile:89]: unnecessary diff

File: src/backend/distributed/shared_library_init.c [hunk 7]
```diff
@@ -354,18 +354,10 @@ _PG_init(void)
...
 	}
-	columnar_init();
+	
+	load_file("columnar.so", false);
 }
 
-
-/* shared library deconstruction function */
-void
-_PG_fini(void)
```
Comment (jeff-davis) [inline src/backend/distributed/shared_library_init.c:363]: Should we remove this from `columnar/mod.c` as well?

File: Makefile [hunk 8]
```diff
@@ -13,10 +13,17 @@ include Makefile.global
 
 all: extension
 
+#build columnar only
+columnar:
+	$(MAKE) -C src/backend/columnar all
+	cp $(citus_abs_srcdir)/src/backend/columnar/columnar.so $(citus_abs_srcdir)/src/backend/distributed/
```
Comment (jeff-davis) [inline Makefile:19]: Maybe there's a better way to use `columnar.so` from where it is, rather than copying it?

File: src/backend/columnar/mod.c [hunk 9]
```diff
@@ -22,16 +22,20 @@
...
 	columnar_tableam_init();
 }
 
-
 void
-columnar_fini(void)
+_PG_fini(void)
 {
 	columnar_tableam_finish();
 }
```
Comment (jeff-davis) [inline src/backend/columnar/mod.c:41]: Let's just get rid of this.

**Comment by jeff-davis** [comment]
It's failing builds in CI:

```
make[1]: Entering directory '/home/circleci/project/build-13/src/backend/columnar'
make[1]: *** No rule to make target 'safeclib/mem_primitives_lib.o', needed by 'columnar.so'.  Stop.
make[1]: Leaving directory '/home/circleci/project/build-13/src/backend/columnar'
```

File: src/backend/columnar/Makefile (hunk 2 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile:24]: Yes, I will remove that

File: src/backend/columnar/Makefile (hunk 3 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile:25]: removed

File: src/backend/columnar/Makefile (hunk 1 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile:20]: done

File: src/backend/columnar/mod.c (hunk 9 above)
Comment (Quicheche) [reply src/backend/columnar/mod.c:41]: done

**Comment by yxu2162** [comment]
> It's failing builds in CI:
> 
> ```
> make[1]: Entering directory '/home/circleci/project/build-13/src/backend/columnar'
> make[1]: *** No rule to make target 'safeclib/mem_primitives_lib.o', needed by 'columnar.so'.  Stop.
> make[1]: Leaving directory '/home/circleci/project/build-13/src/backend/columnar'
> ```

This was the issue I encountered when I changed columnar Makefile to use pathsbst. I think we will need to do further investigation to how distributed finds the safeclib files.

File: src/backend/columnar/Makefile [hunk 10]
```diff
@@ -0,0 +1,22 @@
+safestringlib_srcdir = $(citus_abs_top_srcdir)/vendor/safestringlib
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:1]: Also need a line:

`citus_subdir = src/backend/columnar`

At the top of this Makefile.

File: Makefile (hunk 8 above)
Comment (jeff-davis) [inline Makefile:19]: Should be:

`cp $(citus_top_builddir)/src/backend/columnar/columnar.so $(citus_top_builddir)/src/backend/distributed/`

File: Makefile [hunk 11]
```diff
@@ -13,10 +13,16 @@ include Makefile.global
 
 all: extension
 
+#build columnar only
+columnar:
+	$(MAKE) -C src/backend/columnar all
```
Comment (jeff-davis) [inline Makefile:18]: Above this line, add:

`mkdir -p $(citus_top_builddir)/src/backend/columnar/safeclib`

File: Makefile [hunk 12]
```diff
@@ -13,10 +13,16 @@ include Makefile.global
...
+	cp $(citus_top_builddir)/src/backend/columnar/columnar.so $(citus_top_builddir)/src/backend/distributed/
 # build extension
-extension: $(citus_top_builddir)/src/include/citus_version.h
+extension: $(citus_top_builddir)/src/include/citus_version.h columnar
 	$(MAKE) -C src/backend/distributed/ all
-install-extension: extension
+install-columnar:
+	$(MAKE) -C src/backend/columnar install
+install-extension: extension install-columnar
```
Comment (jeff-davis) [inline Makefile:25]: Also need something under `install-all`

File: src/backend/columnar/Makefile [hunk 13]
```diff
@@ -0,0 +1,22 @@
+citus_subdir = src/backend/columnar
+safestringlib_srcdir = $(citus_abs_top_srcdir)/vendor/safestringlib
+SUBDIRS = . safeclib
+SUBDIRS +=
+OBJS += \
+	$(patsubst $(citus_subdir)/%.c,%.o,$(foreach dir,$(SUBDIRS), $(sort $(wildcard $(citus_subdir)/$(dir)/*.c))))
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:6]: Use `$(citus_abs_srcdir)` instead of `$(citus_subdir)` in both places in this line.

File: src/backend/columnar/Makefile [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+PG_CPPFLAGS += -I$(libpq_srcdir) -I$(safestringlib_srcdir)/include
+
+
+ifdef USE_PGXS
+PG_CONFIG = pg_config
+PGXS := $(shell $(PG_CONFIG) --pgxs)
+include $(PGXS)
+else
+subdir = src/backend/columnar
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:19]: Not needed

File: src/backend/columnar/Makefile [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+top_builddir = ../../..
+include $(top_builddir)/Makefile.global
```
Comment (jeff-davis) [inline src/backend/columnar/Makefile:21]: Use `citus_top_builddir` in both of these lines.

File: src/backend/columnar/mod.c [hunk 16]
```diff
@@ -22,16 +22,14 @@
...
-void
-columnar_init(void)
-{
-	columnar_init_gucs();
-	columnar_tableam_init();
-}
+PG_MODULE_MAGIC;
 
+void _PG_init(void);
+void _PG_fini(void);
```
[the remark this answers is not in the document]
Comment (marcocitus) [reply src/backend/columnar/mod.c]: iirc, _PG_fini is called when someone explicitly does a LOAD 'columnar' when the module is already loaded.

(probably not properly handled by Citus)

File: Makefile (hunk 11 above)
Comment (jeff-davis) [reply Makefile:18]: Oh, looks like it would be better to copy the `ENSURE_SUBDIRS_EXIST` line from `distributed/Makefile`.

File: src/backend/distributed/Makefile [hunk 17]
```diff
@@ -38,10 +36,12 @@ all:
 
 NO_PGXS = 1
 
-SHLIB_LINK = $(libpq)
+SHLIB_LINK += $(libpq) columnar.so
```
[the remark this answers is not in the document]
Comment (Quicheche) [reply src/backend/distributed/Makefile:39]: done

File: src/backend/columnar/Makefile (hunk 13 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile:6]: done

File: src/backend/columnar/Makefile (hunk 14 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile]: removed

File: src/backend/columnar/Makefile (hunk 15 above)
Comment (Quicheche) [reply src/backend/columnar/Makefile]: done

File: Makefile [hunk 18]
```diff
@@ -13,10 +13,16 @@ include Makefile.global
 
 all: extension
 
+#build columnar only
+columnar:
+	$(MAKE) -C src/backend/columnar all
+	cp $(citus_top_builddir)/src/backend/columnar/columnar.so $(citus_top_builddir)/src/backend/distributed/
```
[the remark this answers is not in the document]
Comment (Quicheche) [reply Makefile]: removed

File: Makefile [hunk 19]
```diff
@@ -10,13 +10,21 @@ ifeq (,$(wildcard Makefile.global))
 endif
 
 include Makefile.global
-
+# directories with source files
+SUBDIRS +=  $(citus_top_builddir)/src/backend/columnar/safeclib
+ENSURE_SUBDIRS_EXIST := $(shell mkdir -p $(SUBDIRS))
```
Comment (jeff-davis) [inline Makefile:15]: Should this be in `src/backend/columnar/Makefile` instead? It would seem more symmetrical with `src/backend/distributed/Makefile` and a slightly better place.

File: src/backend/distributed/Makefile [hunk 20]
```diff
@@ -38,10 +36,13 @@ all:
...
 NO_PGXS = 1
 
-SHLIB_LINK = $(libpq)
-
+SHLIB_LINK += $(libpq)
+SHLIB_LINK_INTERNAL += -L../columnar -l:columnar.so
+override enable_rpath = yes
 include $(citus_top_builddir)/Makefile.global
 
+rpathdir = $(shell $(PG_CONFIG) --libdir):$(shell $(PG_CONFIG) --pkglibdir)
```
Comment (jeff-davis) [inline src/backend/distributed/Makefile:44]: Maybe `rpathdir := ...` ?

File: src/backend/distributed/Makefile [hunk 21]
```diff
@@ -38,10 +36,13 @@ all:
 
 NO_PGXS = 1
 
-SHLIB_LINK = $(libpq)
-
+SHLIB_LINK += $(libpq)
```
Comment (jeff-davis) [inline src/backend/distributed/Makefile:39]: I don't think you need `+=` here any more, just `=`.

File: Makefile (hunk 19 above)
Comment (Quicheche) [reply Makefile:15]: but we do columnar "all" here in the top Makefile, and I got below error if I put it under columnar Makefile, " FATAL:  could not load library "/home/yanwjin/test/pg/install.13/lib/postgresql/citus.so": /home/yanwjin/test/pg/install.13/lib/postgresql/citus.so: undefined symbol: ColumnarTableSetOptions_hook",
 Run "ldd citus.so | grep columnar" also returns nothing

File: src/backend/distributed/Makefile (hunk 20 above)
Comment (Quicheche) [reply src/backend/distributed/Makefile:44]: done

File: Makefile (hunk 19 above)
Comment (Quicheche) [reply Makefile:15]: done

File: src/backend/columnar/mod.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (jeff-davis) [reply src/backend/columnar/mod.c]: The docs say it isn't called: https://www.postgresql.org/docs/current/xfunc-c.html

We don't really test that path and so I think we're better off with it gone.

File: Makefile [hunk 22]
```diff
@@ -10,13 +10,18 @@ ifeq (,$(wildcard Makefile.global))
...
-
 all: extension
 
+
+#build columnar only
+columnar:
+	$(MAKE) -C src/backend/columnar all
 # build extension
-extension: $(citus_top_builddir)/src/include/citus_version.h
+extension: $(citus_top_builddir)/src/include/citus_version.h columnar
```
Comment (thanodnl) [inline Makefile:20]: I finally got to tinkering with this PR, first smoketest of a `make install` from the root failed on the linker:
```
Undefined symbols for architecture x86_64:
  "_CheckCitusVersion", referenced from:
      _ColumnarProcessUtility in columnar_tableam.o
      _alter_columnar_table_set in columnar_tableam.o
      _alter_columnar_table_reset in columnar_tableam.o
      _ColumnarTableDropHook in columnar_tableam.o
      _columnar_beginscan in columnar_tableam.o
      _columnar_index_fetch_begin in columnar_tableam.o
      _columnar_tuple_satisfies_snapshot in columnar_tableam.o
      ...
ld: symbol(s) not found for architecture x86_64
Undefined symbols for architecture x86_64:
  "_CheckCitusVersion", referenced from:
      _ColumnarProcessUtility in columnar_tableam.o
      _alter_columnar_table_set in columnar_tableam.o
      _alter_columnar_table_reset in columnar_tableam.o
      _ColumnarTableDropHook in columnar_tableam.o
      _columnar_beginscan in columnar_tableam.o
      _columnar_index_fetch_begin in columnar_tableam.o
      _columnar_tuple_satisfies_snapshot in columnar_tableam.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [columnar.so] Error 1
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [columnar] Error 2
make: *** Waiting for unfinished jobs....
make[1]: *** [columnar.so] Error 1
make: *** [install-columnar] Error 2
```

Is this a known issue? Or am I the first/only to run into this.

I am building on a mac, don't know if @jeff-davis has a similar setup,

When running `make extension` a similar linker error pops up:

```
Undefined symbols for architecture x86_64:
  "_CheckCitusVersion", referenced from:
      _ColumnarProcessUtility in columnar_tableam.o
      _alter_columnar_table_set in columnar_tableam.o
      _alter_columnar_table_reset in columnar_tableam.o
      _ColumnarTableDropHook in columnar_tableam.o
      _columnar_beginscan in columnar_tableam.o
      _columnar_index_fetch_begin in columnar_tableam.o
      _columnar_tuple_satisfies_snapshot in columnar_tableam.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [columnar.so] Error 1
make: *** [columnar] Error 2
```

Let me know if I should debug this further (eg. if I am the only one having this issue)

Comment (jeff-davis) [reply Makefile:20]: Will be fixed when #5688 is merged.

File: src/backend/distributed/shared_library_init.c [hunk 23]
```diff
@@ -346,7 +346,9 @@ _PG_init(void)
 	{
 		DoInitialCleanup();
 	}
-	columnar_init();
+
+
+	load_file("columnar.so", false);
```
Comment (marcocitus) [inline src/backend/distributed/shared_library_init.c:351]: could use a comment, e.g.
```c
/*
 * For convenience and backwards compatibility, we avoid users having to add both
 * citus and columnar to shared_preload_libraries by loading columnar.so as part of
 * loading citus.so.
 */
```

maybe also some compile-time variable for the file name.

Comment (Quicheche) [reply src/backend/distributed/shared_library_init.c:355]: add comment there as suggested

File: Makefile (hunk 22 above)
Comment (jeff-davis) [reply Makefile:20]: I believe this is resolved.

**Review by jeff-davis — APPROVED** [review]

**Review by marcocitus — APPROVED** [review]

File: src/backend/columnar/Makefile [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+	$(patsubst $(citus_abs_srcdir)/%.c,%.o,$(foreach dir,$(SUBDIRS), $(sort $(wildcard $(citus_abs_srcdir)/$(dir)/*.c))))
+
+MODULE_big = citus_columnar
+
+PG_CPPFLAGS += -I$(libpq_srcdir) -I$(safestringlib_srcdir)/include
+
+include $(citus_top_builddir)/Makefile.global
+
+.PHONY: install-all
+install-all: install
```
Comment (marcocitus) [inline src/backend/columnar/Makefile:17]: does this install downgrade scripts too?

Comment (Quicheche) [reply src/backend/columnar/Makefile:17]: No, it won't install downgrade. I ran install-all for my branch vs master, the sql files under share/postgresql/extension are exactly the same, columnar behavior remains same as what we have on current master
